base/gfspvgmgr/virtual_group_manager.go

```diff
@@ -482,3 +504,171 @@ func (vgm *virtualGroupManager) releaseSPAndGVGLoop() {
 		}
 	}
 }
+
+type HealthChecker struct {
+	mutex sync.RWMutex
+
+	sps          map[uint32]*sptypes.StorageProvider // all sps,  id -> StorageProvider
+	unhealthySPs map[uint32]*sptypes.StorageProvider
+
+	chainClient consensus.Consensus // query VG params from chain
+}
+
+func NewHealthChecker(chainClient consensus.Consensus) *HealthChecker {
+	sps := make(map[uint32]*sptypes.StorageProvider)
+	unhealthySPs := make(map[uint32]*sptypes.StorageProvider)
+	return &HealthChecker{sps: sps, unhealthySPs: unhealthySPs, chainClient: chainClient}
+}
+
+func (checker *HealthChecker) addSP(sp *sptypes.StorageProvider) {
+	checker.mutex.Lock()
+	defer checker.mutex.Unlock()
+
+	if !sp.IsInService() {
+		return
+	}
+	if _, exists := checker.sps[sp.GetId()]; !exists {
+		checker.sps[sp.GetId()] = sp
+	}
+}
+
+func (checker *HealthChecker) isSPHealthy(spID uint32) bool {
+	checker.mutex.RLock()
+	defer checker.mutex.RUnlock()
+
+	if sp, exists := checker.sps[spID]; exists {
+		if _, unhealthyExists := checker.unhealthySPs[spID]; unhealthyExists {
+			log.CtxDebugw(context.Background(), "the sp is treated as unhealthy", "sp", sp)
+			return false
+		} else {
+			return true
+		}
+	}
+	log.CtxDebugw(context.Background(), "the sp isn't exist in sps map, is treated as unhealthy", "sps", checker.sps)
+	return false
+}
+
+// isGVGHealthy GVG healthy means gvg primary sp & secondary sps is healthy
+func (checker *HealthChecker) isGVGHealthy(gvg *vgmgr.GlobalVirtualGroupMeta) bool {
+	checker.mutex.RLock()
+	defer checker.mutex.RUnlock()
+
+	// gvg primary sp
+	if !checker.isSPHealthy(gvg.PrimarySPID) {
+		log.CtxDebugw(context.Background(), "the gvg is treated as unhealthy", "gvg", gvg)
+		return false
+	}
+
+	// gvg secondary sp
+	for _, spID := range gvg.SecondarySPIDs {
+		if !checker.isSPHealthy(spID) {
+			log.CtxDebugw(context.Background(), "the gvg is treated as unhealthy", "gvg", gvg)
+			return false
+		}
+	}
+
+	return true
+}
+
+// isVGFHealthy vgf healthy means at least one gvg is healthy
+func (checker *HealthChecker) isVGFHealthy(vgf *vgmgr.VirtualGroupFamilyMeta) bool {
+	checker.mutex.RLock()
+	defer checker.mutex.RUnlock()
+
+	gvgs := vgf.GVGMap
+	if len(gvgs) == 0 {
+		// vgf has no gvg treated as healthy
+		return true
+	}
+
+	healthyCount := len(gvgs)
+	for _, gvg := range gvgs {
+		if !checker.isGVGHealthy(gvg) {
+			healthyCount--
+		}
+	}
+	if healthyCount == 0 {
+		log.CtxDebugw(context.Background(), "the vgf is treated as unhealthy", "vgf", vgf)
+	}
+	return healthyCount > 0
+}
+
+func (checker *HealthChecker) checkAllSPHealth() {
+	spTemp := make(map[uint32]*sptypes.StorageProvider)
+	unhealthyTemp := make(map[uint32]*sptypes.StorageProvider)
+	checker.mutex.RLock()
+	for key, value := range checker.sps {
+		spTemp[key] = value
+	}
+	checker.mutex.RUnlock()
+
+	unhealthyCnt := 0
+	for _, sp := range spTemp {
+		if !checker.checkSPHealth(sp) {
+			unhealthyCnt++
+			unhealthyTemp[sp.GetId()] = sp
+		}
+	}
+
```

Comment:
if unhealthyCnt == 0 {
  // only update checker.sps and return, this is fast-path.
}